blockchain/chainio.go

```diff
@@ -1273,11 +1305,30 @@ func loadBlockIndex(dbTx database.Tx, genesisHash *chainhash.Hash, index *blockI
 			}
 		}
 
-		// Initialize the block node, connect it, and add it to the block
-		// index.
+		// Initialize the block node.
 		node := &blockNodes[i]
 		initBlockNode(node, header, parent)
+
+		// Set the status based on the database entry.
+		//
+		// If new consensus rules were detected, unmark blocks that were
+		// previously marked failed if their median time is after the earliest
+		// possible start time of the new consensus rules.  This ensures clients
+		// that did not update prior to new rules activating are able to
+		// automatically recover under the new rules without having to download
+		// the entire chain again.
```

Comment:
I would suggest documenting the preconditions this logic relies on for correct behavior since it is fairly fragile given it's not really considering the shape of the tree which means it could break assumptions the block index itself makes.

Namely, there are two important preconditions for correct behavior here:

1) The block index entries MUST be loaded in order of their height
2) The median times of blocks are required to be strictly increasing

Reply:
Great suggestion.  The fact that they are loaded in order of their height is part of the reason I had the logic here.  Added a comment for the preconditions.